midend/local_copyprop.cpp

```diff
@@ -419,16 +419,22 @@ IR::MethodCallExpression *DoLocalCopyPropagation::postorder(IR::MethodCallExpres
                 LOG3("table apply method call " << mc->method);
                 apply_table(&tables[obj]);
                 return mc;
-            } else if (mem->expr->type->is<IR::Type_Extern>()) {
+            } else if (mem->expr->type->is<IR::Type_Extern>() ||
+                       mem->expr->type->is<IR::Type_SpecializedCanonical>()) {
                 auto name = obj + '.' + mem->member;
                 if (methods.count(name)) {
                     LOG3("concrete method call " << name);
                     apply_function(&methods[name]);
                     return mc;
                 } else {
-                    LOG3("extern method call " << mc->method << " does nothing");
-                    // FIXME -- is this safe? If an abstract method in outer scope that
-                    // we haven't seen a definition for, it might affects non-locals?
+                    // extern method call might trigger any concrete implementation
+                    // method in the object, so act as if all of them are applied
+                    LOG3("extern method call " << name);
+                    name = obj + '.';
+                    for (auto it = methods.upper_bound(name); it != methods.end(); ++it) {
```

Comment:
I don't think this is conservative enough. The virtual methods could also be invoked asynchronously at any time, e.g., by a request from the control-plane. We have not agreed that abstract methods are only invoked when some other extern method is invoked. 

Also, consider action profiles: the action selector will be invoked when a table is invoked, without any other method being called explicitly. 

So I think you should always consider the abstract methods as "live" and scan them. You know nothing about the context they execute in, though; they could all execute, or one of them could execute multiple times. This code should not be necessary, you should always visit them.